payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -410,11 +410,6 @@ public boolean apply(final PaymentTransactionModelDao input) {
             currentStateName = paymentModelDao.getLastSuccessStateName();
         }
 
-        // Sanity: no paymentMethodId was passed through API and account does not have a default paymentMethodId
-        if (paymentStateContext.getPaymentMethodId() == null) {
-            throw new PaymentApiException(ErrorCode.PAYMENT_NO_DEFAULT_PAYMENT_METHOD, paymentStateContext.getAccount().getId());
```

Comment:
After re-reading the code, it looks like it's safe indeed to remove this check (cannot happen).

However, the issue of trying to pay an invoice (or triggering a standalone payment) without a default payment method can happen, and we should return that specific `ErrorCode` in this case:

* In `JaxRsResourceBase#validatePaymentMethodForAccount`, instead of `verifyNonNull`, can you throw this `ErrorCode`?
* In `DefaultPaymentApi#createPurchaseWithPaymentControl`, instead of `PAYMENT_INVALID_PARAMETER`, we should use that `ErrorCode` instead